CSS Accordion

EdwinGuzman opened the pull request:
Fixes #403 

## **This PR does the following:**
- Updates the `Accordion` component to open and close just with CSS. Instead of using local state and a button, there is now an input element and a label element that apply CSS rules to open/close the content area with display `none` or `inline-block`.

One reason why I wanted this change was that closing the modal unmounted the content area and that caused form values to disappear. So now the content gets rendered all the time.

Some problems with this approach:
* The a11y linter complained that the checkbox did not have the right default value so I had to add `value="empty checkbox"`. This seems unnecessary but the linter complained and now I'm wondering if there'll be any consequences to this. I tried a placeholder which seems less risky but that did not work.
* The a11y linter also complained that it needed the right `aria-*` props and I had to add `aria-checked={false}`. This doesn't seem so bad except that there's no way to update this when the checkbox component _is_ checked since the point of this was trying _not_ to use javascript.
* Lastly, I'm having a hard time testing this with enzyme. It's much easier with jest using something like "check to see if this text string is visible to the user" but that's not available through enzyme. Also, it seems you can't check for CSS rules or DOM element styles like "check if `display` is `none`", etc.

So now, would this be worth continuing and merging?

### Front End Review:
- [ ] View [the example in Storybook]()

EdwinGuzman committed:
First attempt at using just css for the accordion functionality.

EdwinGuzman committed:
Merge branch 'development' into css-accordion

* development:
  Updated changelog for the Input id change.
  Optionally adding and uid=502(edwinguzman) gid=20(staff) groups=20(staff),702(com.apple.sharepoint.group.2),12(everyone),61(localaccounts),79(_appserverusr),80(admin),81(_appserveradm),98(_lpadmin),33(_appstore),100(_lpoperator),204(_developer),250(_analyticsusers),395(com.apple.access_ftp),398(com.apple.access_screensharing),399(com.apple.access_ssh),400(com.apple.access_remote_ae),701(com.apple.sharepoint.group.1) attribute to the Input component.
  Adding the padding and minheight to the correct wrapper and not the Autosuggest example component itself.
  Styling the wrappers for the Autosuggest stories and minor fix to other stories.

EdwinGuzman committed:
Updating Accordion to open/close based on CSS

EdwinGuzman committed:
Removing empty files?

helenvholmes commented:
> * The a11y linter complained that the checkbox did not have the right default value so I had to add `value="empty checkbox"`. This seems unnecessary but the linter complained and now I'm wondering if there'll be any consequences to this. I tried a placeholder which seems less risky but that did not work.
> * The a11y linter also complained that it needed the right `aria-*` props and I had to add `aria-checked={false}`. This doesn't seem so bad except that there's no way to update this when the checkbox component _is_ checked since the point of this was trying _not_ to use javascript.

Quoting this to cc Zach.

zpeyton commented:
Hi @EdwinGuzman and @scooooooooby The linter in this case is a bit overzealous. Standard `<input type="checkbox"` does not need a default value and also do not need aria-* props. Screen readers know how to handle these built in UI components. Only if you make a custom checkbox like, `<div role="checkbox" aria-checked="false">` do you need to use javascript to update the aria attributes with the state change.

EdwinGuzman committed:
Removing strict a11y lint rules and updating tests.

EdwinGuzman commented:
Thanks, @zpeyton .

Updated the PR: removed strict tslint a11y rules and cleaned up the component and test a bit. It's hard to do visual testing with Enzyme and this is one reason why I think we should move to Jest/react-testing-library. Those provide a better way to test components visually.

The test for the `Accordion` is not very in-depth right now but we do have Storybook for interactive UI testing. I tried using `window.getComputedStyle` and it partially worked but did not return the style I was looking for, specifically "none" or "inline-block" for the `display` property.

EdwinGuzman committed:
Merging and fixing conflict.

helenvholmes commented on src/components/Accordion/_Accordion.scss:
```diff
@@ -3,15 +3,35 @@
+    &__label {
         @include button--outline;
 
-        align-items: center;
```
Small nit, but by removing this the icon is no longer centered vertically within the accordion label.

EdwinGuzman replied:
Nice catch. Updated.

helenvholmes reviewed (changes requested):
One small change request, but this looks great—thanks Edwin.

EdwinGuzman committed:
Minor update to align icon in the Accordion component.

helenvholmes commented on src/components/Accordion/_Accordion.scss:
```diff
@@ -3,15 +3,36 @@
-        width: 100%;
+        font-weight: 500;
+        cursor: pointer;
+        align-items: center;
```
Craziest nit but your linter should be bothering you that this should be the first property based on our linting rules—is it not telling you that? https://github.com/NYPL/nypl-design-system/blob/development/.stylelintrc.json#L41

EdwinGuzman replied:
I installed `stylelint` in my vscode so that error shows up now. It looks like stylelint is not setup in the project correctly because that error should come up when running `npm test`.

Although I fixed it for this file, there are errors in other files. This should be a separate ticket not just to fix those files but to make that linter part of the entire process.

I ran `npx stylelint "**/*.scss"` and got this output:

<img width="1337" alt="Screen Shot 2020-12-01 at 1 03 28 PM" src="https://user-images.githubusercontent.com/1280564/100779064-cfa9e280-33d5-11eb-9365-0d20f3b6a5b0.png">

helenvholmes replied:
Oh nice, guess I was only getting those from my editor directly—so lot of stuff was getting missed.

I've filed #450 to track this.

EdwinGuzman committed:
Updating scss based on stylelint.

helenvholmes reviewed (approved):
Looks good!
